skglm/experimental/pdcd_ws.py

```diff
@@ -0,0 +1,225 @@
+import warnings
+
+import numpy as np
+from numpy.linalg import norm
+
+from numba import njit
+from skglm.utils.jit_compilation import compiled_clone
+from sklearn.exceptions import ConvergenceWarning
+
+
+class PDCD_WS:
+    r"""Primal-Dual Coordinate Descent solver with working sets.
+
+    It solves::
+
+        \min_w F(Xw) + G(w) \Leftrightarrow \min_w \max_z <Xw, z> + G(w) - F^*(z)
+
+    where :math:`F` is the datafit term (:math:`F^*` it's Fenchel conjugate)
+    and :math:`G` is the penalty term.
+
+    The datafit is required to be convex and proximable. Also, the penalty
+    is also required to be convex, separable, and proximable.
+
+    The solver is inspired by [1] and uses working sets [2].
+    The working sets are built using a fixed point distance strategy
+    where each feature is assigned a score based how much it maps
+    to itself when performing a primal update::
+
+        \text{score}_j = \abs{w_j - prox_{G_j}(w_j - \tau_j <X_j, z>)}
+
+    where :maths:`\tau_j` is the primal step associated with the j-th feature.
+
+    Parameters
+    ----------
+    max_iter : int, optional
+        The maximum number of iterations or equivalently the
+        the maximum number of solved subproblems.
+
+    max_epochs : int, optional
+        Maximum number of primal CD epochs on each subproblem.
+
+    p0 : int, optional
+        First working set size.
+
+    tol : float, optional
+        The tolerance for the optimization.
+
+    verbose : bool or int, default False
+        Amount of verbosity. 0/False is silent.
+
+    References
+    ----------
+    .. [1] Olivier Fercoq and Pascal Bianchi,
+        "A Coordinate-Descent Primal-Dual Algorithm with Large Step Size and Possibly
+        Nonseparable Functions", SIAM Journal on Optimization, 2020,
+        https://epubs.siam.org/doi/10.1137/18M1168480,
+        code: https://github.com/Badr-MOUFAD/Fercoq-Bianchi-solver
+
+    .. [2] Bertrand, Q. and Klopfenstein, Q. and Bannier, P.-A. and Gidel, G.
+           and Massias, M.
+           "Beyond L1: Faster and Better Sparse Models with skglm", 2022
+           https://arxiv.org/abs/2204.07826
+    """
+
+    def __init__(self, max_iter=1000, max_epochs=1000,
+                 p0=100, tol=1e-6, verbose=False):
+        self.max_iter = max_iter
+        self.max_epochs = max_epochs
+        self.p0 = p0
+        self.tol = tol
+        self.verbose = verbose
+
+    def solve(self, X, y, datafit_, penalty_, w_init=None, Xw_init=None):
+        datafit, penalty = PDCD_WS._validate_init(datafit_, penalty_)
+        n_samples, n_features = X.shape
+
+        # init steps
+        # Despite violating the conditions mentioned in [1]
+        # this choice of steps yield in practice a convergent algorithm
+        # with better speed of convergence
+        dual_step = 1 / norm(X, ord=2)
```

Comment:
this will throw a MemoryError for large sparse matrices as it casts X to dense

Reply:
I didn't handle the sparse case

Reply:
This may be going again the principle of least astonishment as other solvers do. Do you throw a ValueError when X is sparse ?